edgehog-device-runtime-containers/src/docker/volume.rs

```diff
@@ -0,0 +1,336 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Handles the container Volumes.
+
+use std::{
+    collections::HashMap,
+    fmt::{Debug, Display},
+};
+
+use bollard::{
+    errors::Error as BollardError, models::Volume as DockerVolume, volume::CreateVolumeOptions,
+};
+use tracing::{debug, error, instrument, trace, warn};
+
+use crate::client::*;
+
+/// Error for the image operations.
+#[non_exhaustive]
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+pub enum VolumeError {
+    /// couldn't create the volume
+    Create(#[source] BollardError),
+    /// couldn't inspect volume
+    Inspect(#[source] BollardError),
+    /// couldn't complete volume operation, volume in use
+    InUse(#[source] BollardError),
+    /// couldn't remove volume
+    Remove(#[source] BollardError),
+    /// couldn't list volumes
+    List(#[source] BollardError),
+}
+
+/// Docker volume struct.
+///
+/// Persistent storage that can be attached to containers.
+#[derive(Debug, Clone, Eq)]
+pub struct Volume<S> {
+    /// The volume's name. If not specified (empty), Docker generates a name.
+    pub name: S,
+    /// Name of the volume driver to use.
+    ///
+    /// Defaults to "local".
+    pub driver: S,
+    /// A mapping of driver options and values.
+    ///
+    /// These options are passed directly to the driver and are driver specific.
+    pub driver_opts: HashMap<String, S>,
+}
+
+impl<S> Volume<S> {
+    /// Create a new volume.
+    pub fn new(name: S, driver: S, driver_opts: HashMap<String, S>) -> Self {
+        Self {
+            name,
+            driver,
+            driver_opts,
+        }
+    }
+
+    /// Create a new docker volume.
+    ///
+    /// See the [Docker API reference](https://docs.docker.com/engine/api/v1.43/#tag/Volume/operation/VolumeCreate)
+    #[instrument]
+    pub async fn create(&self, client: &Client) -> Result<(), VolumeError>
+    where
+        S: Debug + Display + AsRef<str>,
+    {
+        debug!("Create the Volume {}", self);
+
+        client
+            .create_volume(self.into())
+            .await
+            .map_err(VolumeError::Create)?;
+
+        Ok(())
+    }
+
+    /// Inspect a docker volume.
+    ///
+    /// See the [Docker API reference](https://docs.docker.com/engine/api/v1.43/#tag/Volume/operation/VolumeInspect)
+    #[instrument]
+    pub async fn inspect(&self, client: &Client) -> Result<Option<DockerVolume>, VolumeError>
+    where
+        S: Debug + Display + AsRef<str>,
+    {
+        debug!("inspecting volume {}", self.name);
+
+        let res = client.inspect_volume(self.name.as_ref()).await;
+
+        match res {
+            Ok(volume) => {
+                trace!("volume info: {volume:?}");
+
+                Ok(Some(volume))
+            }
+            Err(BollardError::DockerResponseServerError {
+                status_code: 404,
+                message,
+            }) => {
+                warn!("volume not found: {message}");
+
+                return Ok(None);
```

Comment:
You can remove `return`.
```suggestion
                Ok(None)
```

Reply:
Thanks fixed